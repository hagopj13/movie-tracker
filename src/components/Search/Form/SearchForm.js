// @flow
import React, { useEffect } from 'react';
import { useForm } from 'react-hook-form';
import * as yup from 'yup';

import SearchField from 'components/Search/Field/SearchField';

type Props = {
  selectedQuery: string,
  onSubmit: ({ query: string }) => void,
};

const schema = yup.object().shape({
  query: yup.string().min(1).required(),
});

const SearchForm = (props: Props) => {
  const { selectedQuery, onSubmit } = props;

  const { register, handleSubmit, setValue } = useForm({
    mode: 'onSubmit',
    validationSchema: schema,
  });

  useEffect(() => {
    if (selectedQuery) {
      setValue('query', selectedQuery);
    }
  }, [selectedQuery, setValue]);

  return (
    <form onSubmit={handleSubmit(onSubmit)}>
      <SearchField name="query" inputRef={register} />
    </form>
  );
};

export default SearchForm;
