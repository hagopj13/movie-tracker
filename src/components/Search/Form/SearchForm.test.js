import React from 'react';
import { act } from 'react-dom/test-utils';
import { createShallow, createMount } from '@material-ui/core/test-utils';

import SearchForm from './SearchForm';

describe('SearchForm component', () => {
  let shallow;
  let mount;

  beforeAll(() => {
    shallow = createShallow();
    mount = createMount();
  });

  it('should render the SearchForm component correctly', () => {
    const selectedQuery = 'Some query';
    const mockProps = {
      selectedQuery,
      onSubmit: jest.fn(),
    };
    const wrapper = shallow(<SearchForm {...mockProps} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should change the SearchField value when the selected query changes', () => {
    const selectedQuery = 'Some query';
    const mockProps = {
      selectedQuery,
      onSubmit: jest.fn(),
    };
    const wrapper = mount(<SearchForm {...mockProps} />);
    const searchElement = wrapper.find('input');
    expect(searchElement.getDOMNode().value).toBe(selectedQuery);

    const newQuery = 'Some new query';
    wrapper.setProps({ selectedQuery: newQuery });
    expect(searchElement.getDOMNode().value).toBe(newQuery);
  });

  it('should call onSubmit when the form is submitted', async () => {
    const mockSubmit = jest.fn();
    const mockProps = {
      selectedQuery: '',
      onSubmit: mockSubmit,
    };
    const wrapper = mount(<SearchForm {...mockProps} />);

    const query = 'Some query';
    await act(async () => {
      const searchElement = wrapper.find('input');
      searchElement.getDOMNode().value = query;
      searchElement.getDOMNode().dispatchEvent(new Event('input'));

      wrapper.find('form').simulate('submit');
    });

    expect(mockSubmit).toHaveBeenCalledTimes(1);
    expect(mockSubmit.mock.calls[0][0]).toEqual({ query });
  });

  it('should not call onSubmit when the form is submitted with an empty query', async () => {
    const mockSubmit = jest.fn();
    const mockProps = {
      selectedQuery: 'Some query',
      onSubmit: mockSubmit,
    };
    const wrapper = mount(<SearchForm {...mockProps} />);

    const searchElement = wrapper.find('input');
    searchElement.getDOMNode().value = '';
    searchElement.getDOMNode().dispatchEvent(new Event('input'));

    await act(async () => {
      wrapper.find('form').simulate('submit');
    });

    expect(mockSubmit).not.toHaveBeenCalled();
  });
});
