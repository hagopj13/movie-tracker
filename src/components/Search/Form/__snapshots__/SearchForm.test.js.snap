// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchForm component should render the SearchForm component correctly 1`] = `
<form
  onSubmit={[Function]}
>
  <SearchField
    inputRef={[Function]}
    name="query"
  />
</form>
`;
